docs/reference/gadget-helper-api.md

```diff
@@ -0,0 +1,110 @@
+---
+title: 'Gadget helper API'
+weight: 10
+description: 'Reference documentation for the gadget helper API'
+---
+
+The gadget helper API is a set of features exposed by Inspektor Gadget accessible from the eBPF code.
+
+## Container enrichment
+
+To make use of container enrichment, gadgets must include the relevent header:
+```
+#include <gadget/mntns_filter.h>
+```
+
+Inspektor Gadget will automatically enrich events with container information
+when the events include the mount namespace in this way:
+
+```
+struct event {
+        gadget_mntns_id mntns_id;
+        /* other fields */
+}
+```
+
+An eBPF program can look up the mount namespace with the following:
+```
+u64 mntns_id;
+mntns_id = gadget_get_mntns_id();
+```
+
+EBPF programs of type socket filter cannot use `gadget_get_mntns_id()` but can
+use socket enrichment to find the mount namespace.
+
+## Container filtering
+
+EBPF programs should stop processing an event when it does not originate from a
+container of interest.
+
+```
+if (gadget_should_discard_mntns_id(mntns_id))
+        return 0;
+```
+
+## Socket enrichment
+
+To make use of socket enrichment, gadgets must include the relevent header.
+
+For eBPF programs of type socket filter:
+```
+#define GADGET_TYPE_NETWORKING
+#include <gadget/sockets-map.h>
+```
+
+For eBPF programs of other kinds:
+```
+#define GADGET_TYPE_TRACING
+#include <gadget/sockets-map.h>
+```
+
+This will define the following struct:
+```
+#define TASK_COMM_LEN 16
+struct sockets_value {
+        __u64 mntns;
+        __u64 pid_tgid;
+        __u64 uid_gid;
+        char task[TASK_COMM_LEN];
+        /* other private fields */
+};
+```
+
+Then, an eBPF can find additional details about a socket with
+`gadget_socket_lookup()`.
+
+For eBPF programs of type socket filter:
+
+```
+struct sockets_value *skb_val = gadget_socket_lookup(skb);
+if (skb_val != NULL) {
+        /* Access skb_val->mntns and other fields */
```

Comment:
Thanks for the suggested plan, I think it makes sense for a guide. Let me describe my thoughts about this documentation: I wanted to separate the guides from the reference documentation.

Guides:
* Explain the developer how to do a specific task. The task might involve using several different APIs.
* The guide should follow the flow of a developer task: describing each task a developer should do to develop their gadget.
* Only give the recommended way, not deprecated information
* Examples:
  * https://inspektor-gadget.io/docs/latest/devel/hello-world-gadget/

Reference documentation:
* Exhaustive list of an API currently implemented
* A reference documentation should list the API in some kind of order (alphabetic or by category) but it is not a chronological order of tasks
* Include compatibility information for previous versions of the API
* Is useful if someone wants to find out if they can do a new task that has not been explained in a guide
* Is useful if someone wants to write an alternative implementation of Inspektor Gadget to run existing gadgets

With that in mind, I think that some of your listed items should be added in a guide "how to do enrichment in your gadget".

And I think the enriched types should be part of `gadget-helper-api.md`  because it is part of the same API, even if they are here to achieve a different goal (so they can be described in a different guide too).

Reply:
Oh, that makes perfectly sense for me. I kind of mixed the two concepts.